src/take.rs

```diff
@@ -0,0 +1,52 @@
+use alloc::vec::Vec;
+use core::mem;
+
+use crate::{Archetype, DynamicBundle, TypeInfo};
+
+/// An entity removed from a `World`
+pub struct TakenEntity<'a> {
+    archetype: &'a Archetype,
```

Comment:
I think a `&'a mut Archetype` would be better documenting the underlying assumptions here. It already is essential that as `TakenEntity<'a>` is produced by a `&'a mut World` and I don't think this is straight foward to generalize to taken multiple entities as there is only one "one past the end" slot per archetype.

Reply:
Since this basically has exclusive access to `World` and thereby `Archetype`, I think it could also be more efficient by avoiding the preparatory swap with the last entity stored in the source archetype? (Thereby going from 3+1 copies to 1+1, i.e. entity to temp, last entity to entity, temp to last entity and last entity to destination versus entity to destination and last entity to entity.)

Reply:
> I think a &'a mut Archetype would be better documenting the underlying assumptions here

I agree. Unfortunately, a mutable reference can't be moved out of `Self`, which means we can't `mem::forget(self)` before moving components out of `archetype` in `put`, which means a panic inside `put`'s callback might cause `Drop` to be invoked on values which have already been moved away, which is UB. We could work around this by storing the reference in `Option`, but that consumes the niche, which seems unfortunate.

> it could also be more efficient by avoiding the preparatory swap with the last entity stored in the source archetype?

The current design is motivated by maintaining good behavior even if `TakenEntity` is leaked. However, on review I think we can still preserve that by deferring world changes to `put`. I'm kind of curious exactly how much worse `swap_nonoverlapping` is than `copy_nonoverlapping`, though, since avoiding the former is what this proposal boils down to.

Reply:
Revised. The result is a slightly larger `TakenEntity`, but simpler code overall, which I'm happy with.